### ISSUE TITLE
I still get links errors despite #4978 landing

### ISSUE THREAD

**Issue by Eh2406** [rust-lang/crates.io#7310]
I maintain a `-sys` crate but my users keep getting 
```
error: multiple packages link to native library `z`, but a native library can be linked only once
```
I thought that had been fixed in rust-lang/cargo#4978 but it doesn't seem to be working. What can I do?

--------------------

In fact this is mostly a place for me to link people to, to unify the discussion. I will try to keep this up to date with any discussion below.

> it doesn't seem to be working

Here is why rust-lang/cargo#4978 doesn't seem to be working for you.
rust-lang/cargo#4978 only stops that error for crate with the `links` attribute in the [index](https://github.com/rust-lang/crates.io-index). That attribute only gets into the index if you publish the crate with a recent nightly/beta. So no protection for the older versions of your crate.

> What can I do?

You can manually make a PR to the [index](https://github.com/rust-lang/crates.io-index) to add the attribute to older versions.

Someone could scan all crates to find all the missing `links` attributes and add them.

It is not yet clear what the cargo team recommends at this time.

**Comment by Turbo87** [comment]
I'll keep it in mind for the next round of backfilling, but I agree that it is probably low priority :)

**Comment by Turbo87** [comment]
FWIW the original issue has been fixed. https://github.com/rust-lang/crates.io-index/blob/master/li/bz/libz-sys now includes `links: z` for the relevant crate versions. once a new version is published we automatically sync the whole index file for the crate with the database which fixes the issue. 

if you maintain a `-sys` crate and experience a similar issue: publish a new version and then the issue should be fixed for the older versions as well.

**Comment by epage** [comment]
I'm a bit confused by that explanation.  To double check my understanding: when a new package is published, you re-parse all previous manifests and write their index entries from scratch?  That seems quite expensive which is why I suspect my understanding is incorrect.  If that isn't it, how does publishing a new version help existing index entries?

**Comment by Turbo87** [comment]
> when a new package is published, you re-parse all previous manifests and write their index entries from scratch?

no, when a new package is published we parse the manifest and save the relevant stuff in our database. the database is then used to regenerate the index file for all existing versions, not just the newly published version. this allowed us to get rid of various inconsistencies in the index last year that had snuck in over time.

**Comment by Turbo87** [comment]
the database was backfilled from the git index (see https://github.com/rust-lang/crates.io/pull/5112). we were previously storing a subset of the data in the index and a subset in the database. with the linked PR and some of the related ones we converted the database to be the source of truth and changed the index to be derived from the data in the database.

**Comment by epage** [comment]
The question posed was about backfilling from `Cargo.toml`, like we did with `package.rust-version`, and not from the database or the git index.

Like I said, enough time has passed that this might not be worth fixing (or was already fixed).  I just want to make sure we are on the same page on the problem and closing this out for the right reasons.

**Comment by Turbo87** [comment]
ah, I see what you mean. the issue was in the index, we backfilled the database from the index, so the issue got into the database as well and so creating the index files from the database doesn't help... 😅

**Comment by Turbo87** [comment]
alright, **now** this should actually be fixed: https://github.com/rust-lang/crates.io-index/commit/c2ea799f8ef79ad3cc9422d41a1a9a4716b152e6

for more context: https://rust-lang.zulipchat.com/#narrow/stream/318791-t-crates-io/topic/backfilling.20.60links.60.20fields
